### PR TITLE
initial

### PR DESCRIPTION
**Pull request by ilgooz**

### CODE REVIEW

**Review by ilgooz — COMMENTED** [review]
webapp needs small improvements in the typescript side, some types aren't correct and compiler config does not complain about it.

on the server side, maybe we can also use `errors.Wrap()` to create more oblivious errors like in the other plugins.

**Review by ilgooz — COMMENTED** [review]
add url versioning prefix to HTTP API: `/api/v1/`

**Comment by ilgooz** [comment]
add CI steps as described here: https://developers.mattermost.com/extend/plugins/community_process/#ci-system-setup

**Review by ilgooz — COMMENTED** [review]
add CI steps as described here: https://developers.mattermost.com/extend/plugins/community_process/#ci-system-setup

File: README.md [hunk 1]
```diff
@@ -0,0 +1,28 @@
+## mattermost-plugin-topdf
+_TOPDF_ plugin converts Office files that attached to posts to PDF files and caches them for future use. To convert a file to PDF simply make a `GET` request to:
```
Comment (lieut-data) [inline README.md]: The instructions to convert a file here seem out of scope -- I'd suggest making the language focus on the webapp preview vs. the API used internally to achieve same. I'd also suggest lifting up the Gotenburg dependency:

```suggestion
_TOPDF_ plugin automatically generates PDF previews of Microsoft Office files attached to posts for use with Mattermost. The plugin itself depends on [Gotenburg](https://github.com/thecodingmachine/gotenberg) for the heavy lifting, but seamlessly intercepts attachments to generate the previews without further user interaction.
```

File: README.md [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+```
+http://{mattermostAddr}/plugins/topdf/files/{fileID}
+```
+
+This plugin also enables previewing Office files on _webapp_.
```
Comment (lieut-data) [inline README.md]: I think this can be eliminated with the proposed changes above.

File: README.md [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+## Installation
+1. Go to the [releases page of this Github repository](https://github.com/ilgooz/mattermost-plugin-topdf/releases) and download the latest release for your Mattermost server.
+   
+2. In the Mattermost System Console under **System Console > Plugins > Plugin Management** upload the file to install the plugin. To learn more about how to upload a plugin, [see the documentation](https://docs.mattermost.com/administration/plugins.html#plugin-uploads).
+
+3. Install _Gotenberg_ server since this plugin depends on it to actually convert files to PDFs. Assuming you have already installed Docker, following command will download and install the latest version of _Gotenberg_ and set up a server with an open port at 4798:
```
Comment (lieut-data) [inline README.md]: Would it make sense to just link out to https://thecodingmachine.github.io/gotenberg/#install instead of replicating those steps here? Also, should this be step 1. to ensure the dependencies are met before going on to next steps?

File: README.md [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+  ```
+    docker run -d -p 4798:3000 --env DEFAULT_WAIT_TIMEOUT=600 --env MAXIMUM_WAIT_TIMEOUT=600 thecodingmachine/gotenberg:6
+  ```
+  Make sure that you've set large timeouts for Gotenberg server like in the sample above since converting big files takes time.
+
+4. Once _Gotenberg_ server is running, configure the plugin to make requests to your _Gotenberg_ instance. Go to **System Console > Plugins > TOPDF** and configure **Gotenberg's Full Address** to point at your _Gotenberg_ instance.  
+
+5. Activate the plugin at **System Console > Plugins > Management** and ensure it starts with no errors.
```
Comment (lieut-data) [inline README.md]: It's actually possible to enable the plugin right from within the `System Console > Plugins > TOPDF` page as well.

File: plugin.json [hunk 5]
```diff
@@ -0,0 +1,36 @@
+{
+  "id": "topdf",
+  "name": "TOPDF",
+  "description": "This plugin converts Office files that attached to posts to PDF files and enables previewing for Office files on `webapp`",
```
Comment (lieut-data) [inline plugin.json]:
```suggestion
  "description": "Automatically generate PDF previews of Microsoft Office files attached to posts using Gotenburg.",
```

File: plugin.json [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+  },
+  "server": {
+    "executables": {
+        "linux-amd64": "server/dist/plugin-linux-amd64",
+        "darwin-amd64": "server/dist/plugin-darwin-amd64",
+        "windows-amd64": "server/dist/plugin-windows-amd64.exe"
+    }
+  },
+  "settings_schema": {
+    "header": "PDF plugin to convert uploaded files to pdf. Ony Office files supported at this moment.",
```
Comment (lieut-data) [inline plugin.json]:
```suggestion
    "header": "PDF plugin to convert uploaded files to pdf. Only Office files supported at this moment.",
```

File: plugin.json [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+    "footer": "",
+    "settings": [{
+      "key": "GotenbergAddress",
+      "display_name": "Gotenberg's Full Address",
+      "type": "text",
+      "help_text": "This plugin uses Gotenberg server to convert files to PDFs. See [documentation here](https://thecodingmachine.github.io/gotenberg).\n\n **warning!** don't forget to set proper timeouts as you need in Gotenberg server, for ex:\n `docker run -d -p 4798:3000 --env DEFAULT_WAIT_TIMEOUT=600 --env MAXIMUM_WAIT_TIMEOUT=600 thecodingmachine/gotenberg:6`",
```
Comment (lieut-data) [inline plugin.json:24]: Can you elaborate here on the timeouts? Is that a concern specific to this plugin, or something general for when configuring Gotenberg?

File: plugin.json [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+      "placeholder": "http://localhost:4798",
+      "default": "http://localhost:4798"
+    },{
+      "key": "GotenbergConvertTimeout",
+      "display_name": "File Convert Timeout",
+      "type": "text",
+      "help_text": "This timeout set while initializing a convert request to Gotenberg server and while waiting for whole response to be finished. See timeout format [here](https://golang.org/pkg/time/#ParseDuration).",
```
Comment (lieut-data) [inline plugin.json:31]: Traditionally, our configuration options just use seconds vs expecting a trailing `s` or parsing other options. Might be good to leave it that way for consistency.

File: server/gotenberg/gotenberg.go [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+	"net/url"
+	"sync"
+	"time"
+)
+
+const (
+	// pingTimeout set while sending ping requests to Gotenberg server.
+	pingTimeout = time.Second * 5
+
+	// defaultConvertTimeout set while sending and waiting response ofr file convert requests
```
Comment (lieut-data) [inline server/gotenberg/gotenberg.go]:
```suggestion
	// defaultConvertTimeout defines the timeout for file conversion requests to the Gotenberg server
```

File: server/gotenberg/gotenberg.go [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+	"io/ioutil"
+	"mime/multipart"
+	"net/http"
+	"net/url"
+	"sync"
+	"time"
+)
+
+const (
+	// pingTimeout set while sending ping requests to Gotenberg server.
```
Comment (lieut-data) [inline server/gotenberg/gotenberg.go]:
```suggestion
	// pingTimeout defines the timeout for ping requests to the Gotenberg server.
```

File: server/gotenberg/gotenberg.go [hunk 11]
```diff
@@ -0,0 +1,197 @@
...
+func (g *Gotenberg) cloneConfing() config {
+	g.m.RLock()
+	defer g.m.RUnlock()
+	return *g.conf
+}
+
+// Option used to customize Gotenberg defaults.
+type Option func(*Gotenberg)
+
+// ConvertTimeoutOption customizes default convert timeout set while sending file convert requests
```
Comment (lieut-data) [inline server/gotenberg/gotenberg.go]: Suggest perhaps naming this `ConversionTimeoutOption`:
```suggestion
// ConversionTimeoutOption sets the timeout for conversion requests.
```

File: server/gotenberg/gotenberg.go [hunk 12]
```diff
@@ -0,0 +1,197 @@
...
+// to Gotenberg server.
+func ConvertTimeoutOption(convertTimeout time.Duration) Option {
+	return func(g *Gotenberg) {
+		g.conf.convertTimeout = convertTimeout
+	}
+}
+
+// AddrOption updates addr set during call to New().
```
Comment (lieut-data) [inline server/gotenberg/gotenberg.go]:
```
// AddrOption changes the configured Gotenberg server address.
```

File: server/gotenberg/gotenberg.go [hunk 13]
```diff
@@ -0,0 +1,197 @@
...
+
+// New creates new Gotenberg client with given Gotenberg server addr and options.
+func New(addr string, options ...Option) *Gotenberg {
+	g := &Gotenberg{conf: &config{addr: addr}}
+	g.UpdateConfig(options...)
+	return g
+}
+
+// UpdateConfig updates options set during the first initialization of New.
+func (g *Gotenberg) UpdateConfig(options ...Option) {
```
Comment (lieut-data) [inline server/gotenberg/gotenberg.go]: Being able to change the configuration seems to come with the complexity of having to safeguard the config from concurrent access. What if we just forced the client to create a new `Gotenberg` instance in that case instead?

File: server/gotenberg/gotenberg.go [hunk 14]
```diff
@@ -0,0 +1,197 @@
...
+	}
+	// check if Gotenberg is cool with the file we sent to see if it's gonna response back with a PDF data.
+	if res.StatusCode != http.StatusOK &&
+		res.StatusCode != http.StatusCreated {
+		defer res.Body.Close()
+		// request is somehow not successful:
+		// file content can be invalid or some timeout might be hitting set by Gotenberg's end or here in the request.
+		data, err := ioutil.ReadAll(res.Body)
+		if err != nil {
+			return nil, err
```
Comment (lieut-data) [inline server/gotenberg/gotenberg.go]: suggest `error.Wrap` here to add context to this particular failure point.

File: server/gotenberg/gotenberg.go [hunk 15]
```diff
@@ -0,0 +1,197 @@
...
+		}
+		return nil, fmt.Errorf("error from Gotenberg with '%d' code: %s", res.StatusCode, string(data))
+	}
+	// we have Gotenberg willing to stream PDF data, give it to the caller so it can start reading.
+	return res.Body, nil
+}
+
+// buildGotenbergURL generates a Gotenberg API URL from given addr for endpoint.
+// TODO(ilgooz): maybe cache generated URLs in memory to improve performance.
```
Comment (lieut-data) [inline server/gotenberg/gotenberg.go]: Can you elaborate on the potential performance concern here?

File: server/gotenberg/gotenberg_test.go [hunk 16]
```diff
@@ -0,0 +1,62 @@
...
+	}))
+	defer ts.Close()
+	gt := New(ts.URL)
+	isRunning, err := gt.Status()
+	require.NoError(t, err)
+	require.True(t, isRunning)
+}
+
+func TestStatusNotRunning(t *testing.T) {
+	gt := New("http://not-existent-host")
```
Comment (lieut-data) [inline server/gotenberg/gotenberg_test.go]: Suggest not making a DNS request out to `not-existent-host` -- could we mock an `httptest` server that instead doesn't respond to the ping?

File: server/plugin.go [hunk 17]
```diff
@@ -0,0 +1,147 @@
...
+	// app is the actual, underlying TOPDF app and its features exposed to
+	// network via Plugin's HTTP API.
+	app interface {
+		CheckServerStatus() (running bool, err error)
+		GetPDF(userID, fileID string) (pdf io.ReadCloser, err error)
+	} // *topdf.TOPDF
+
+	// gt used by the app to interact with topdf.PDFServer implementation(Gotenberg).
+	// it's also used by Plugin to update Gotenberg's configs when Plugin configs reloaded.
+	gt *gotenberg.Gotenberg
```
Comment (lieut-data) [inline server/plugin.go]: Given the (nice) abstraction between the plugin and the app, it seems unfortunate to have to expose the Gotenberg layer at all here. Should we instead surface the necessary API on the `TOPDF` struct instead?

File: server/plugin.go [hunk 18]
```diff
@@ -0,0 +1,147 @@
...
+func (p *Plugin) OnConfigurationChange() error {
+	var conf configuration
+	if err := p.API.LoadPluginConfiguration(&conf); err != nil {
+		return err
+	}
+	options := []gotenberg.Option{
+		gotenberg.AddrOption(conf.GotenbergAddress),
+		gotenberg.ConvertTimeoutOption(time.Duration(conf.GotenbergConvertTimeout)),
+	}
+	p.gt.UpdateConfig(options...)
```
Comment (lieut-data) [inline server/plugin.go]: See above, but it seems unfortunate to have to "know" about Gotenberg at this layer -- maybe we just proxy the configuration to the application instead, and let it reconfigure?

File: server/plugin.go [hunk 19]
```diff
@@ -0,0 +1,147 @@
...
+	}
+	defer pdf.Close()
+	w.Header().Set("Content-Type", "application/pdf")
+	// stream PDF content to requester.
+	io.Copy(w, pdf)
+}
+
+// logError logs errors with Plugin API.
+func (p *Plugin) logError(err error) {
+	p.API.LogError(err.Error())
```
Comment (lieut-data) [inline server/plugin.go:119]: I believe `p.API.LogError` can now just take any `error` as of v5.16 or v5.18 -- it literally does `err.Error()` to transmit as a string across the wire -- so it might be reasonable to eliminate this helper.

File: server/topdf/topdf.go [hunk 20]
```diff
@@ -0,0 +1,143 @@
...
+// toPDFPrefix used as a prefix while using KV store to access PDF files' fileID.
+const toPDFPrefix = "pdf:"
+
+// ErrUnauthorizedUser returned when user has no access to a file that requested to be converted to PDF.
+var ErrUnauthorizedUser = errors.New("user is not authorized to access pdf")
+
+// TOPDF is an application that converts files to PDFs and permanently caches them by using Mattermost APIs.
+type TOPDF struct {
+	// mapi is Mattermost's Plugin API.
+	mapi plugin.API
```
Comment (lieut-data) [inline server/topdf/topdf.go:23]: Accepting the full `plugin.API` here seems to somewhat defeat the point of factoring out `topdf` as a server. Should we instead accept an interface with just the required functionality?

File: server/topdf/topdf.go [hunk 21]
```diff
@@ -0,0 +1,143 @@
...
+	// if there is no PDF file cached, create it, cache and use its content.
+	if len(pid) == 0 {
+		data, err := t.createAndSavePDF(fileID, userID)
+		if err != nil {
+			return nil, err
+		}
+		return ioutil.NopCloser(bytes.NewReader(data)), nil
+	}
+	// we have the PDF version in cache, directly return it back.
+	data, aerr := t.mapi.GetFile(string(pid))
```
Comment (lieut-data) [inline server/topdf/topdf.go]: We should verify the user has access to the file here as well, otherwise we will serve up the PDF after someone with access generates it.

File: server/topdf/topdf.go [hunk 22]
```diff
@@ -0,0 +1,143 @@
...
+	if _, aerr := t.mapi.GetChannelMember(post.ChannelId, userID); aerr != nil {
+		return nil, normalizeAppErr(aerr)
+	}
+	// get file's content by fileID.
+	byts, aerr := t.mapi.GetFile(fileID)
+	if aerr != nil {
+		return nil, normalizeAppErr(aerr)
+	}
+	// convert file to PDF by using PDF server.
+	r, err := t.server.Convert(info.Name, info.Extension, bytes.NewReader(byts))
```
Comment (lieut-data) [inline server/topdf/topdf.go]: Would it make sense to leverage your distributed lock here to avoid creating it more than once, as well as to allow other plugin requests to block until the file is created? Can we use the file id as the key?

File: server/topdf/topdf.go [hunk 23]
```diff
@@ -0,0 +1,143 @@
...
+	post, aerr := t.mapi.GetPost(info.PostId)
+	if aerr != nil {
+		return nil, normalizeAppErr(aerr)
+	}
+	// check if the user has access to the channel where associated post submitted.
+	if _, aerr := t.mapi.GetChannelMember(post.ChannelId, userID); aerr != nil {
+		return nil, normalizeAppErr(aerr)
+	}
+	// get file's content by fileID.
+	byts, aerr := t.mapi.GetFile(fileID)
```
Comment (lieut-data) [inline server/topdf/topdf.go]: nit: `byts` isn't an idiomatic abbreviation. Perhaps `fileBytes`?

File: server/x/doc.go [hunk 24]
```diff
@@ -0,0 +1,2 @@
+// Packages x contains xpackages where each package's purpose to extend features of its origin.
+package x
```
Comment (lieut-data) [inline server/x/doc.go:2]: An interesting pattern!

File: webapp/src/delete/i18n/i18n.jsx [hunk 25]
```diff
@@ -0,0 +1,196 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
```
Comment (lieut-data) [inline webapp/src/delete/i18n/i18n.jsx]: I'm assuming this is all to support the extracted component? I'd suggest dropping all the localization and assume we expose the component in some way in the future instead, since there are a lot more strings here than needed, and we wouldn't be able to maintain these translations consistently anyway.

**Review by lieut-data — CHANGES_REQUESTED** [review]
Super cool, @ilgooz! I've gone through the code in detail and provided some feedback. Nice work :)

File: README.md (hunk 2 above)
Comment (ilgooz) [reply README.md]: removed by b16f1afa1fef546a7dcce06e8df55c0472e5b539

File: README.md (hunk 3 above)
Comment (ilgooz) [reply README.md]: I wanted to point out that Gotenberg server needs custom timeouts since creating PDFs might take some time and defaults are too low.

I also dropped a Docker command there that includes correct timeouts so this way I think it's a bit easier for users to just copy-paste that command and run the server.

But I'll add the installation link too.

done by b16f1afa1fef546a7dcce06e8df55c0472e5b539 and 78c470413a966f1ddd00f8a0d0b165d7d138379a

File: README.md (hunk 4 above)
Comment (ilgooz) [reply README.md]: updated by b16f1afa1fef546a7dcce06e8df55c0472e5b539

File: plugin.json (hunk 7 above)
Comment (ilgooz) [reply plugin.json:24]: This is a concern while configuring Gotenberg. Its timeouts are too low and we generally hit them while converting mind-size files. I wanted to get an extra attention here.

What do you think, should we keep this here or move somewhere else like header?

File: server/gotenberg/gotenberg.go (hunk 13 above)
Comment (ilgooz) [reply server/gotenberg/gotenberg.go]: Right. We have a concurrency safe implementation anyway but I agree that we can just get rid of `UpdateConfig` and reinitialize the app on every configuration change.

It'll make things even simpler. Also, this is still will be efficient, since configuration updates shouldn't happen very often.

P.S.: I wanted to experiment a new thing that different than what we have in other plugins when I come up with _UpdateConfig_. But it's even better to take reinitialization approach.

File: server/gotenberg/gotenberg.go (hunk 15 above)
Comment (ilgooz) [reply server/gotenberg/gotenberg.go]: Nothing much really. It would just save us from some additional cpu cycles. Nothing mandatory, I'll remove this TODO.

File: server/gotenberg/gotenberg_test.go (hunk 16 above)
Comment (ilgooz) [reply server/gotenberg/gotenberg_test.go]: Right, nice catch, thanks! I'll just change it to `http://localhost`.

It's probably not needed to use `httptest` here since in this test, we need to test the behavior when Gotenberg server is not running at all.

--edit

Tho, I should refactor `Status()` to check for a `200` response. In that case spinning up an `httptest` would be a better solution since this will let us to run this test in much isolated environment because we don't know what's running at `localhost:80`.

File: server/plugin.go (hunk 17 above)
Comment (ilgooz) [reply server/plugin.go]: I was liking this approach because it's easier to mock Gotenberg's funcs this way while testing `topdf` pkg without the need of faking in a lower, `httptest` layer.

--

Otherwise, if we still want to mock Gotenberg's funcs, we need to create multiple interfaces around.

For example, `gotenberg.New()` has to return an interface instead of `*gotenberg.Gotenberg` and `topdf.New()` has to accept an interface with `gotenberg.New()` func in it.

I kind of don't like taking this approach since it feels a bit not idiomatic to me.

--

In the `main` plugin package, I think of `*gotenberg.Gotenberg` like a dependency to `*topdf.TOPDF` and it feels very natural to me.

Also this way, it is very easy to configure dependencies without the need of proxying configurations to the dependency via it's dependent, in this case `topdf.TOPDF`.

I remember that I first saw this technique in go-micro a long time ago and adopted since.

What do you think how we should proceed on this one?

--

P.S. from [go-micro](https://micro.mu/docs/framework.html#plugin-option):
```go
package main

import (
        "github.com/micro/go-micro" 
        // consul registry
        "github.com/micro/go-plugins/registry/consul"
        // rabbitmq transport
        "github.com/micro/go-plugins/transport/rabbitmq"
        // kafka broker
        "github.com/micro/go-plugins/broker/kafka"
)

func main() {
	registry := consul.NewRegistry()
	broker := kafka.NewBroker()
	transport := rabbimq.NewTransport()

        service := micro.NewService(
                micro.Name("greeter"),
                micro.Registry(registry),
                micro.Broker(broker),
                micro.Transport(transport),
        )

	service.Init()
	service.Run()
}
```

File: server/plugin.go (hunk 18 above)
Comment (ilgooz) [reply server/plugin.go]: related to https://github.com/ilgooz/mattermost-plugin-topdf/pull/1#discussion_r346544207 and https://github.com/ilgooz/mattermost-plugin-topdf/pull/1#discussion_r346548396

File: server/topdf/topdf.go (hunk 20 above)
Comment (ilgooz) [reply server/topdf/topdf.go:23]: I partially agree that it defeats the point of moving this logic to a separate package because as you said it still depends on the `plugin.API ` instead of a much more generic interface.

But still in my view, this abstraction builds some solid lines between the `main` pkg and actual `topdf` logic.

On the other hand, it's kind of hard to create a much more generic interface other than directly using `plugin.API` for `topdf` to accept because topdf heavily depends on `plugin.API`'s features.

That being said, I can still create an interface to only select the funcs that we need use from `plugin.API` as you pointed out but still this interface will be strongly tied to `plugin.API` since it'll be direct copy of its some funcs. But in a general perspective, having this interface seems better to be able to see what Plugin APIs that `topdf` package depends at the first look.

To get more precise about the design pattern I follow here,

I consider `plugin.API` as a `DAO` layer/provider here. `topdf` pkg is the actually app that uses DAO and I place the `main` pkg as a user interface layer to expose topdf's features to the network layer by using the hooks including ServeHTTP (-REST API handlers).

As a summary in the architectural point of view:
* `topdf` pkg is the actual app(controller) and uses `plugin.API` DAO to interact with an external API, in this case Mattermost server's features.

* `main` pkg is the network interface to expose features of `topdf` to other 3rd parties and also to configure `topdf` pkg to match with the needs of those 3rd parties.

File: server/topdf/topdf.go (hunk 21 above)
Comment (ilgooz) [reply server/topdf/topdf.go]: Ops, my bad. I'll be fixing this one!

File: server/topdf/topdf.go (hunk 22 above)
Comment (ilgooz) [reply server/topdf/topdf.go]: Yes, yes!

File: server/x/doc.go (hunk 24 above)
Comment (ilgooz) [reply server/x/doc.go:2]: Thank you! I hope us to bring a @mattermost/x repo!

File: webapp/src/delete/i18n/i18n.jsx (hunk 25 above)
Comment (ilgooz) [reply webapp/src/delete/i18n/i18n.jsx]: Agreed!

Comment (ilgooz) [reply webapp/src/delete/i18n/i18n.jsx]: removed by 5938c68602c6d7452063c377277bd7ad44edde6c

Comment (ilgooz) [reply webapp/src/delete/i18n/i18n.jsx]: removed by 9a871503e422a8d52e2ceb0293728c112714b373

File: server/gotenberg/gotenberg.go (hunk 12 above)
Comment (ilgooz) [reply server/gotenberg/gotenberg.go]: this option is completely removed by 3e03bd479501bd15845b7b0793bd601acffda8b2

File: server/gotenberg/gotenberg.go (hunk 13 above)
Comment (ilgooz) [reply server/gotenberg/gotenberg.go]: removed by 3e03bd479501bd15845b7b0793bd601acffda8b2

File: server/gotenberg/gotenberg.go (hunk 12 above)
Comment (ilgooz) [reply server/gotenberg/gotenberg.go]: this option is completely removed by 3e03bd4

File: server/topdf/topdf.go (hunk 23 above)
Comment (ilgooz) [reply server/topdf/topdf.go]: done by 3a0f014

File: server/gotenberg/gotenberg.go (hunk 15 above)
Comment (ilgooz) [reply server/gotenberg/gotenberg.go]: done by 3a0f014

File: server/gotenberg/gotenberg.go (hunk 14 above)
Comment (ilgooz) [reply server/gotenberg/gotenberg.go]: done by 3a0f014

File: server/gotenberg/gotenberg_test.go (hunk 16 above)
Comment (ilgooz) [reply server/gotenberg/gotenberg_test.go]: solved by 68f3a3951a40b7bb7c75285dfbac2e0e2185e602

File: server/plugin.go (hunk 19 above)
Comment (ilgooz) [reply server/plugin.go:119]: That's great! Is it this version?: https://github.com/mattermost/mattermost-server/blob/release-5.18/plugin/api.go#L849

File: server/topdf/topdf.go (hunk 21 above)
Comment (ilgooz) [reply server/topdf/topdf.go]: fixed by 0e022692a902a9e071b77837e3f187b66d393e6d
